### DISCUSSION TITLE
Supported camera

### DISCUSSION THREAD

**Discussion by vanDerAlgiers** [ros-industrial-consortium/scan_n_plan_workshop#8] [Q&A]
Hello,

Is it possible to use FANUC 3D camera with this package? If not what are the supported cameras?

Regards,

**Comment by marip8** [comment]
The only requirement is that your camera publish depth and color images in ROS format (`sensor_msgs::Image`) and its depth camera intrinsic parameters in the form of a `sensor_msgs::CameraInfo` message. I'm not sure if there is a ROS driver for the Fanuc 3D camera, but if there is I think it would work

**Comment by vanDerAlgiers** [comment]
Hello @marip8, 

Thank you for your reply.
The camera returns a PCD file. Is it possible to use it directly or a driver should be developed to go from the PCD data to sensor_msgs?

Regards

**Reply by gavanderhoorn** [reply]
> The camera returns a PCD file

is that a regular, PointCloud library PCD, or a proprietary file format which Fanuc happens to have given the same `.pcd` extension?

**Reply by vanDerAlgiers** [reply]
It is a .pcd file.

**Reply by gavanderhoorn** [reply]
But is the data contained by it an actual [PointCloudLibrary](https://pointclouds.org/) PCD?

**Reply by marip8** [reply]
Assuming the output of the camera is a point cloud, you won't be able to use the reconstruction pipeline that we implemented in this project. However, you could write your own node that satisfies the `open3d_interface_msgs::srv::StartYakReconstruction` and `open3d_interface_msgs::srv::StopYakReconstruction` ROS interfaces and produces a triangle mesh file (e.g. `.stl`, `.ply`) from the point cloud from your sensor. There are tools in Open3D, PCL and [`industrial_pcl`](https://github.com/ros-industrial/industrial_pcl) that can do this type of surface reconstruction

**Reply by vanDerAlgiers** [reply]
> But is the data contained by it an actual [PointCloudLibrary](https://pointclouds.org/) PCD?

The file data is as follow:

```
\# .PCD v0.7 - Point Cloud Data file format
VERSION 0.7
FIELDS x y z
SIZE 4 4 4
TYPE F F F
COUNT 1 1 1 
WIDTH 8037
HEIGHT 1
VIEWPOINT 0 0 0 1 0 0 0
POINTS 8037
DATA binary
458.171 1516.28 321.981
460.747 1516.34 321.893
463.302 1516.41 321.429
465.858 1516.48 320.965
468.426 1516.55 320.69
470.996 1516.61 320.414
473.573 1516.67 320.263
476.145 1516.74 319.988
478.705 1516.81 319.462
481.269 1516.88 318.998
483.854 1516.94 318.91
486.424 1517.01 318.509
489.015 1517.06 318.547
491.59 1517.13 318.209
494.166 1517.2 317.87
...
```

**Reply by gavanderhoorn** [reply]
Nice. I would not have expected that.

**Reply by vanDerAlgiers** [reply]
> Nice. I would not have expected that.

However, the raw file data is in binary form, but I converted it to human readable format.

**Comment by marip8** [comment]
> However, you could write your own node that satisfies the open3d_interface_msgs::srv::StartYakReconstruction and open3d_interface_msgs::srv::StopYakReconstruction ROS interfaces and produces a triangle mesh file (e.g. .stl, .ply) from the point cloud from your sensor

These service type I mentioned above will soon change to [the services defined in this new repo](https://github.com/ros-industrial/industrial_reconstruction/tree/main/industrial_reconstruction_msgs/srv)

**Reply by vanDerAlgiers** [reply]
So the straight way to use this project is to use Kinect?

**Reply by marip8** [reply]
Yes that would be the most straightforward way. Any camera that publishes RGB and depth images (Kinect, Realsense, Astra, etc.) should work. We used a Realsense D455
